hypothesis-python/src/hypothesis/database.py

```diff
@@ -324,3 +335,287 @@ def delete(self, key: bytes, value: bytes) -> None:
+                )
+        except URLError:
+            warning_message = "Could not connect to GitHub to get the latest artifact. "
+        except TimeoutError:
+            warning_message = "Could not connect to GitHub to get the latest artifact (connection timed out). "
+
+        if warning_message:
+            warnings.warn(HypothesisWarning(warning_message))
+            return None
+
+        # Get the latest artifact from the list
+        artifact = sorted(
+            filter(lambda a: a["name"] == self.artifact_name, artifacts),
+            key=lambda a: a["created_at"],
+        )[-1]
+
+        request = Request(
+            artifact["archive_download_url"],
+            headers={
+                "Accept": "application/vnd.github+json",
+                "X-GitHub-Api-Version": "2022-11-28",
+                "Authorization": f"Bearer {self.token}",
+            },
+        )
+
+        # Download the artifact
+        try:
+            response = urlopen(request)
+            artifact_bytes = response.read()
+            warning_message = None
+        except HTTPError as e:
+            if e.code == 401:
+                warning_message = (
+                    "Authorization failed when trying to download artifact from GitHub. "
+                    "Check your $GITHUB_TOKEN environment variable or make the repository public. "
+                )
+            else:
+                warning_message = (
+                    "Could not get the latest artifact from GitHub. "
+                    "This could be because because the repository or artifact does not exist. "
+                )
+        except URLError:
+            warning_message = "Could not connect to GitHub to get the latest artifact. "
+        except TimeoutError:
+            warning_message = "Could not connect to GitHub to get the latest artifact (connection timed out). "
+        if warning_message is not None:
+            warnings.warn(HypothesisWarning(warning_message))
+            return None
+
+        # Save the artifact to the cache
+        artifact_path = self.path / f"{datetime.datetime.now().isoformat()}.zip"
+        try:
+            with open(artifact_path, "wb") as f:
+                f.write(artifact_bytes)
+        except OSError:
+            warnings.warn(
+                HypothesisWarning("Could not save the latest artifact from GitHub. ")
+            )
+            return None
+
+        # Test that the artifact is valid
+        try:
+            with ZipFile(artifact_path) as f:
+                if f.testzip():
+                    raise BadZipFile
+        except BadZipFile:
+            warnings.warn(
+                HypothesisWarning(
+                    "The downloaded artifact from GitHub is invalid. "
+                    "This could be because the artifact was corrupted, "
+                    "or because the artifact was not created by Hypothesis. "
+                )
+            )
+            return None
```

Comment:
I think that this should probably go in the initialize method.  Currently, we'll leave the invalid archive on disk but _not_ warn about it when it's already been downloaded.  We may also want to delete it and fall back to an older archive?